service/controller/resource/updateinfrarefs/create.go

```diff
@@ -0,0 +1,94 @@
+package updateinfrarefs
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/service/controller/controllercontext"
+	"github.com/giantswarm/cluster-operator/service/controller/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := meta.Accessor(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	or, err := r.toObjRef(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Here we fetch the provider specific CR defined as infrastructure reference
+	// in the CAPI type. We use an unstructured object and therefore need to set
+	// the api version and kind accordingly. If we would not do that the
+	// controller-runtime client cannot find the right object.
+	ir := &unstructured.Unstructured{}
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding infrastructure reference")
+
+		ir.SetAPIVersion(or.APIVersion)
+		ir.SetKind(or.Kind)
+
+		err = r.k8sClient.CtrlClient().Get(ctx, key.ObjRefToNamespacedName(or), ir)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found infrastructure reference")
+	}
+
+	var updated bool
+
+	// Syncing the provider operator version label, e.g. for aws-operator,
+	// kvm-operator or the like.
+	{
+		l := fmt.Sprintf("%s-operator.giantswarm.io/version", r.provider)
```

Comment:
This is fine for now, but we might need to turn this into loop that iterates over all matches for `^[^-]*-operator.giantswarm.io\/version$` labels so that the whole release keeps coherent.

Reply:
Right, I figured this could be more generic but is good enough for now. More version labels than the one of the provider operator have to be addressed when we make more operators part of the release.